crud/batch_insert.lua

```diff
@@ -0,0 +1,185 @@
+    dev_checks('string', 'table', {
+        timeout = '?number',
+        fields = '?table',
+    })
+
+    opts = opts or {}
+
+    local space = utils.get_space(space_name, vshard.router.routeall())
+    if space == nil then
+        return nil, {BatchInsertError:new("Space %q doesn't exist", space_name)}, true
+    end
+
+    local batches_by_replicasets, err = sharding.split_tuples_by_replicaset(tuples, space)
+    if err ~= nil then
+        return nil, {err}, true
+    end
+
+    local batch_insert_on_storage_opts = {
+        fields = opts.fields,
+    }
+
+    local call_opts = {
+        timeout = opts.timeout,
+        is_async = true,
+    }
+
+    local futures_by_replicasets = {}
+    for replicaset, batch in pairs(batches_by_replicasets) do
+        local func_args = {
+            space_name,
+            batch,
+            batch_insert_on_storage_opts,
+        }
+
+        local future = replicaset:call(BATCH_INSERT_FUNC_NAME, func_args, call_opts)
+        futures_by_replicasets[replicaset.uuid] = future
+    end
+
+    local results, errs = call.batch(
+        futures_by_replicasets,
+        BATCH_INSERT_FUNC_NAME,
+        opts.timeout
+    )
+
+    local rows = {}
+    for _, result in pairs(results) do
+        rows = utils.table_extend(rows, result[1])
+    end
+
+    if next(rows) == nil then
+        return nil, errs
+    end
+
+    local res, err = utils.format_result(rows, space, opts.fields)
+    if err ~= nil then
+        return nil, {err}
+    end
+
+    return res, errs
+end
+
+--- Batch inserts tuples to the specified space
+--
+-- @function tuples_batch
+--
+-- @param string space_name
+--  A space name
+--
+-- @param table tuples
+--  Tuples
+--
+-- @tparam ?table opts
+--  Options of batch_insert.tuples_batch
+--
+-- @return[1] tuples
+-- @treturn[2] nil
+-- @treturn[2] table of tables Error description
+
+function batch_insert.tuples_batch(space_name, tuples, opts)
+    checks('string', 'table', {
+        timeout = '?number',
+        fields = '?table',
+    })
+
+    return schema.wrap_func_reload(call_batch_insert_on_router, space_name, tuples, opts)
+end
+
+--- Batch inserts objects to the specified space
+--
+-- @function objects_batch
+--
+-- @param string space_name
+--  A space name
+--
+-- @param table objs
+--  Objects
+--
+-- @tparam ?table opts
+--  Options of batch_insert.tuples_batch
+--
+-- @return[1] objects
+-- @treturn[2] nil
+-- @treturn[2] table of tables Error description
+
+function batch_insert.objects_batch(space_name, objs, opts)
+    checks('string', 'table', {
+        timeout = '?number',
+        fields = '?table',
+    })
+
+    local tuples = {}
+    for _, obj in ipairs(objs) do
+
+        local tuple, err = utils.flatten_obj_reload(space_name, obj)
+        if err ~= nil then
+            local err_obj = BatchInsertError:new("Failed to flatten object: %s", err)
+            err_obj.tuple = obj
+            return nil, {err_obj}
+        end
```

Comment:
https://github.com/tarantool/crud/pull/232#discussion_r877425419

About `object` as error field: I think it would be a little bit confusing that in some cases we get error with `tuple` field in case there is an error on storage when we call `*_object_many` but also we can get an error with `object` field in flatten object error case.

Reply:
It still doubts me, to be honest.

If I'll read the README I have no chance to learn that the `tuple` field can contain something that is not a tuple. Since this field may be used for programmatic decisions (say, what to retry), its format should be explicitly and unambiguously specified.

So either add another field for an object (and add it to the README) or clarify current behaviour in the README. I would prefer the former.

There is the third option: use a neural field name. Say, `operation_data`.